Doc note update for complex autograd

anjali411 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45270 Doc note update for complex autograd**

<img width="1679" alt="Screen Shot 2020-10-07 at 1 45 59 PM" src="https://user-images.githubusercontent.com/20081078/95368324-fa7b2d00-08a3-11eb-9066-2e659a4085a2.png">
<img width="1673" alt="Screen Shot 2020-10-07 at 1 46 10 PM" src="https://user-images.githubusercontent.com/20081078/95368332-fbac5a00-08a3-11eb-9be5-77ce6deb8967.png">
<img width="1667" alt="Screen Shot 2020-10-07 at 1 46 30 PM" src="https://user-images.githubusercontent.com/20081078/95368337-fe0eb400-08a3-11eb-80a2-5ad23feeeb83.png">
<img width="1679" alt="Screen Shot 2020-10-07 at 1 46 48 PM" src="https://user-images.githubusercontent.com/20081078/95368345-00710e00-08a4-11eb-96d9-e2d544554a4b.png">
<img width="1680" alt="Screen Shot 2020-10-07 at 1 47 03 PM" src="https://user-images.githubusercontent.com/20081078/95368350-023ad180-08a4-11eb-89b3-f079480741f4.png">
<img width="1680" alt="Screen Shot 2020-10-07 at 1 47 12 PM" src="https://user-images.githubusercontent.com/20081078/95368364-0535c200-08a4-11eb-82fc-9435a046e4ca.png">

Differential Revision: [D24203257](https://our.internmc.facebook.com/intern/diff/D24203257)

anjali411 committed:
Doc note update for complex numbers

[ghstack-poisoned]

anjali411 committed:
Update on "Doc note update for complex autograd"




[ghstack-poisoned]

anjali411 commented:
cc. @boeddeker

albanD commented on docs/source/notes/autograd.rst:
```diff
@@ -219,24 +219,40 @@ Autograd for Complex Numbers
+            return s
 
-where :math:`1j` is a unit imaginary number.
+where :math:`1j` is a unit imaginary number and `u` and `v` are :math:`ℝ^{2} → ℝ^{2}` functions.
```
u and v are R2 -> R functions here no?

anjali411 replied:
right my bad

albanD commented on docs/source/notes/autograd.rst:
```diff
@@ -219,24 +219,40 @@ Autograd for Complex Numbers
+
+are Conjugate Wirtinger and Wirtinger derivatives respectively.
+
+For `grad_out = c+dj \in C`, the :math:`VJP` can also be expressed as:
```
nit space in c+dj

albanD commented on docs/source/notes/autograd.rst:
```diff
@@ -245,49 +261,42 @@ where
-
-We define the :math:`VJP` of :math:`F` at :math:`(x, y)` for a cotangent vector :math:`c+dj \in C` as:
+The key assumption made in the derivation of the above formula is that function :math:`F` is a part of a
+real valued function. For more details, please check out Section 3.4 `here <https://arxiv.org/pdf/1701.00392.pdf>`_.
```
I would maybe add another sentence here before the link to be a bit more precise like:
"It implies that the Conjugate Wirtinger and Wirtinger derivative flowing back are complex conjugate of each other. This allows us to use `grad_output` as the Conjugate Wirtinger derivative and recover the Wirtinger derivative by complex conjugation."

anjali411 replied:
gotcha! I made an update based on your suggestion. I specified that here we are talking about the incoming `Conjugate Wirtinger derivative` and `Wirtinger derivative` to ensure the reader doesn't confuse them with the Wirtinger derivatives for `F`

albanD commented on docs/source/notes/autograd.rst:
```diff
@@ -245,49 +261,42 @@ where
 
-Based on formulas above and the behavior we expect to see (going from :math:`ℂ → ℝ^2 → ℂ` should be an identity),
-we use the formula given below for cross-domain functions.
+Note that the Wirtinger and Conjugate Wirtinger derivative in this case are conjugate of each other.
```
I am not sure that this note is important here and it might be confusing with the point we make above when phrased this way.

anjali411 replied:
point taken and removed!

albanD commented on docs/source/notes/autograd.rst:
```diff
@@ -245,49 +261,42 @@ where
-    .. math:: VJP = \begin{bmatrix} c & d \end{bmatrix} * J * \begin{bmatrix} 1 \\ -1j \end{bmatrix}
+        def F(z):
+            x = real(z)     # z = real(z)
+            s = u(x, y) + v(x, y) * 1j
```
You need to define y to be able to write this.
Should it be y=0?

albanD commented on docs/source/notes/autograd.rst:
```diff
@@ -245,49 +261,42 @@ where
+    .. math:: real(grad_out^{*} * \frac{\partial s}{\partial z^{*}} + grad_out * (\frac{\partial s}{\partial z^{*}})^{*})
 
-    .. math:: VJP = \begin{bmatrix} c & -d \end{bmatrix} * J
+Note that the gradient in this case equals to the real value of the result obtained using the above formula. This is because
```
You might want to be more precise about "above formula" :D

anjali411 replied:
lol I agree. added labels to make it more clear!

albanD commented on docs/source/notes/autograd.rst:
```diff
@@ -245,49 +261,42 @@ where
 
-    .. math:: VJP = \begin{bmatrix} c & -d \end{bmatrix} * J
+Note that the gradient in this case equals to the real value of the result obtained using the above formula. This is because
+in the derivation of the formula earlier, we assumed that the function in question is :math:`F: ℂ → ℂ`. However, by redoing the math
```
C -> R ?

anjali411 replied:
I am talking about the F used above (which was `C -> C`) which itself is a part of bigger `C->R` (which has been stated as well).

albanD reviewed (commented):
Looks mostly good for me.
Just small comments.

anjali411 committed:
Update on "Doc note update for complex autograd"




[ghstack-poisoned]

anjali411 committed:
Update on "Doc note update for complex autograd"




[ghstack-poisoned]

anjali411 committed:
Update on "Doc note update for complex autograd"




[ghstack-poisoned]

anjali411 committed:
Update on "Doc note update for complex autograd"




[ghstack-poisoned]

anjali411 committed:
Update on "Doc note update for complex autograd"




[ghstack-poisoned]

zou3519 commented on docs/source/notes/autograd.rst:
```diff
@@ -219,24 +219,44 @@ Autograd for Complex Numbers
-PyTorch follows `JAX's <https://jax.readthedocs.io/en/latest/notebooks/autodiff_cookbook.html#Complex-numbers-and-differentiation>`_
-convention for autograd for Complex Numbers.
+PyTorch uses the Wirtinger Calculus to compute gradients for functions with complex valued input and (or)
+output. PyTorch's autograd convention for complex numbers allows gradient value to be used in
```
nit: "gradient values to be used"

zou3519 commented on docs/source/notes/autograd.rst:
```diff
@@ -245,49 +265,51 @@ where
-We define the :math:`VJP` of :math:`F` at :math:`(x, y)` for a cotangent vector :math:`c+dj \in C` as:
+The key assumption made in the derivation of :eq:`[1]` is that function :math:`F` is a part of a
+real valued function. It implies that the Conjugate Wirtinger and Wirtinger derivative flowing back
+to :math:`F` are complex conjugate of each other. This allows us to use :math:`grad\_output` as the incoming Conjugate
```
nit: "are complex conjugates" (plural)

zou3519 commented on docs/source/notes/autograd.rst:
```diff
@@ -245,49 +265,51 @@ where
+            return s
 
-For any other desired behavior, you can specify the covector `grad_output` in :func:`torch.autograd.backward` call accordingly.
+at :math:`x + yj` can be simplified to be written as:
```
nit: "can be simplified to be written as" -> "can be simplified as"

zou3519 commented on docs/source/notes/autograd.rst:
```diff
@@ -219,24 +219,44 @@ Autograd for Complex Numbers
+We define the Vector-Jacobian Product `VJP` of :math:`F` at :math:`x + yj` for grad output vector :math:`grad\_out` as:
 
-    .. math:: \begin{bmatrix} 1 & 1j \end{bmatrix} * J * \begin{bmatrix} c \\ d \end{bmatrix}
+    .. math::
+        VJP = grad\_out^{*} * \frac{\partial s}{\partial z^{*}} + grad\_out * (\frac{\partial s}{\partial z})^{*}
+        :label: [1]
```
nit: It's a bit hard to read through latex, doc changes are easier to review if you provide a screenshot.

zou3519 commented on docs/source/notes/autograd.rst:
```diff
@@ -219,24 +219,44 @@ Autograd for Complex Numbers
             x, y = real(z), imag(z)
-            return u(x, y) + v(x, y) * 1j
+            s = u(x, y) + v(x, y) * 1j
+            return s
 
-where :math:`1j` is a unit imaginary number.
+where :math:`1j` is a unit imaginary number and :math:`u` and :math:`v` are :math:`ℝ^{2} → ℝ` functions.
 
-We define the :math:`JVP` for function :math:`F` at :math:`(x, y)` applied to a tangent
-vector :math:`c+dj \in C` as:
+We define the Vector-Jacobian Product `VJP` of :math:`F` at :math:`x + yj` for grad output vector :math:`grad\_out` as:
```
As a complex numbers n00b, I think this documentation is a great step to helping developers understand how to write (or update) gradient formulas to support complex numbers.

I think something else that would be helpful is if there was an Appendix somewhere that stepped through the derivation of some operator (like torch.mul, or torch.dot). When I started out understanding how complex number gradients worked, I tried to derive some formulas (e.g., for torch.mul) and compare them against the existing gradient formulas.

Perhaps that would be better for developer documentation though -- I'm not sure if the purpose of notes/autograd is for users to look at and understand what's going on, or for developers to read.

anjali411 replied:
@zou3519 thanks for the feedback! I added the screenshots in the description as per your suggestion. I agree that it would be useful to step through the derivation for one of the operators. However, I am not sure if this doc note is the right place to do that. We can possibly include a gradient derivation for one of the operators here: https://pytorch.org/docs/stable/complex_numbers.html#autograd along with an example of how the gradients can be used in an optimization problem.

anjali411 committed:
Update on "Doc note update for complex autograd"




[ghstack-poisoned]

boeddeker commented on docs/source/notes/autograd.rst:
```diff
@@ -245,49 +265,51 @@ where
-mode automatic differentiation. Notice that d and :math:`1j` are negated in the formula above. Please look at
-the `JAX docs <https://jax.readthedocs.io/en/latest/notebooks/autodiff_cookbook.html#Complex-numbers-and-differentiation>`_
-to get explanation for the negative signs in the formula.
+The above formula can also be verified to be correct for cross-domain functions functions.
```
Typo: One `functions` too much in `functions functions`.

boeddeker commented on docs/source/notes/autograd.rst:
```diff
@@ -219,24 +219,44 @@ Autograd for Complex Numbers
             x, y = real(z), imag(z)
-            return u(x, y) + v(x, y) * 1j
+            s = u(x, y) + v(x, y) * 1j
+            return s
```
Just an idea. You start here directly with the assumption, that the chain rule is known.
Maybe it would be better for a beginner, when you start with the core assumption:
e.g.: Interpreted the complex number as two numbers and calculate the derivative:

 grad_x = dG / dx where G \in \mathcal{R}

 grad_y = dG / dy where G \in \mathcal{R}

Since complex numbers can contain 2 real numbers, we can define:

grad = grad_x + 1j * grad_y

or

grad = grad_x - 1j * grad_y

And this is known as Wirtinger Calculus.

(Note: Dropping the 1/2 in eq. (3.2) of https://arxiv.org/pdf/1701.00392.pdf is usually what NNs do. At the time of writing the paper we weren't careful, because this equation is never implemented)

boeddeker commented on docs/source/notes/autograd.rst:
```diff
@@ -245,49 +265,51 @@ where
+            return s
 
-The :math:`JVP` and :math:`VJP` for a :math:`f1: ℝ^2 → ℂ` are defined as:
+at :math:`x + yj` can be simplified to:
```
This is always difficult to describe. But you cannot evaluate it at x + yj, because z is real valued.
Or should this operation be `C -> R -> C -> R` i.e. `G(F(real_to_complex(real(z)))))` (Not sure, if this matches your notation. Probably is G in your case the complete function)  ?

ezyang commented on docs/source/notes/autograd.rst:
```diff
@@ -219,24 +219,44 @@ Autograd for Complex Numbers
+PyTorch uses the Wirtinger Calculus to compute gradients for functions with complex valued input and (or)
+output. PyTorch's autograd convention for complex numbers allows gradient values to be used in
+Gradient Descent with the same update rule as that in the case of real numbers. This convention works for
+optimization problems where the objective function is real valued.
```
It probably would still be helpful to compare with JAX and Tensorflow here (we match TF, and are the opposite of JAX). It also seems like a good idea to tell people (1) that there is a choice AT ALL to be made here, and (2) compare and contrast the benefits of each convention.

anjali411 committed:
Update on "Doc note update for complex autograd"


<img width="1672" alt="Screen Shot 2020-09-24 at 1 09 05 PM" src="https://user-images.githubusercontent.com/20081078/94177334-5f2b9600-fe67-11ea-97cf-b2949153a117.png">
<img width="1673" alt="Screen Shot 2020-09-24 at 1 09 16 PM" src="https://user-images.githubusercontent.com/20081078/94177346-62bf1d00-fe67-11ea-94cc-d429b55a73b1.png">


[ghstack-poisoned]

anjali411 committed:
Update on "Doc note update for complex autograd"


<img width="1672" alt="Screen Shot 2020-09-24 at 1 09 05 PM" src="https://user-images.githubusercontent.com/20081078/94177334-5f2b9600-fe67-11ea-97cf-b2949153a117.png">
<img width="1673" alt="Screen Shot 2020-09-24 at 1 09 16 PM" src="https://user-images.githubusercontent.com/20081078/94177346-62bf1d00-fe67-11ea-94cc-d429b55a73b1.png">


[ghstack-poisoned]

anjali411 commented:
@boeddeker thanks so much for your feedback. I incorporated your suggestions of walking through some basic steps of what gradients for complex numbers look like by explaining gradient calculations for multiplication operation. let me know what you think!

@ezyang I updated the doc based on our last discussion. I am less inclined to add Tensorflow JAX comparison because:
1. Tensorflow doesn't have documentation explaining their behavior for complex functions. 
2. Our result is not always conjugate (or "opposite") to that of JAX. We actually get completely different results when grad_out is set to a value other than 1.0 or even if a bunch ops line up (and the net effect of conjugation introduced through tf formulas becomes null).

This is how the doc renders:
<img width="1680" alt="Screen Shot 2020-10-01 at 11 45 55 PM" src="https://user-images.githubusercontent.com/20081078/94886330-d30a0780-0440-11eb-801d-926a64288b21.png">
<img width="1677" alt="Screen Shot 2020-10-01 at 11 46 19 PM" src="https://user-images.githubusercontent.com/20081078/94886336-d9987f00-0440-11eb-858c-d8a3f9365100.png">
<img width="1675" alt="Screen Shot 2020-10-01 at 11 46 58 PM" src="https://user-images.githubusercontent.com/20081078/94886351-e1f0ba00-0440-11eb-8ebd-db036bc505c8.png">
<img width="1679" alt="Screen Shot 2020-10-01 at 11 47 11 PM" src="https://user-images.githubusercontent.com/20081078/94886355-e3ba7d80-0440-11eb-81f6-cd5053e624e8.png">

boeddeker commented on docs/source/notes/autograd.rst:
```diff
@@ -219,24 +219,58 @@ Autograd for Complex Numbers
+PyTorch uses the Wirtinger Calculus to compute gradients for functions with complex valued input and (or)
+output. Wirtinger Calculus provides a means of computing gradients of real valued cost functions defined on
+complex domains(:math:`ℂ^n`) which are not holomorphic (complex differentiable). This comes in very handy since
+most of the commonly used linear algebraic functions are not holomorphic.
```
I am not sure, if the statement
 `... most of the commonly used linear algebraic functions are not holomorphic.`
will be correctly interpreted from a user.
Actually many complex operations are holomorphic.
The drawback is, that there is no relevant holomorphic function to end up with a cost function (i.e. fulfill the Cauchy-Riemann equations).

boeddeker commented on docs/source/notes/autograd.rst:
```diff
@@ -245,49 +279,133 @@ where
+from the last `ℂ -> ℝ` function in :math:`G`, say :math:`H`, are complex conjugates of each other.
+This allows us to use :math:`grad\_output` as the incoming Conjugate Wirtinger derivative and recover the
+incoming Wirtinger derivative by complex conjugation of :math:`grad\_output` for function :math:`H`.
+For more details, please check out Section 3.4 `here <https://arxiv.org/pdf/1701.00392.pdf>`_.
```
Is this easier to understand, when you define 
`dJ/ds* = dJ/du + dJ/dv j` and `dJ/ds = dJ/du - dJ/dv j`.
Then it is obvious, that `dJ/ds* = (dJ/ds)*`.

boeddeker commented:
For the example of the complex derivation, it would be nice to extend the example with PyTorch code.
Depending on your intention, either some code to verfy your example or a custom function with a manual defined gradient.

anjali411 committed:
Update on "Doc note update for complex autograd"


<img width="1672" alt="Screen Shot 2020-09-24 at 1 09 05 PM" src="https://user-images.githubusercontent.com/20081078/94177334-5f2b9600-fe67-11ea-97cf-b2949153a117.png">
<img width="1673" alt="Screen Shot 2020-09-24 at 1 09 16 PM" src="https://user-images.githubusercontent.com/20081078/94177346-62bf1d00-fe67-11ea-94cc-d429b55a73b1.png">


[ghstack-poisoned]

anjali411 commented:
> For the example of the complex derivation, it would be nice to extend the example with PyTorch code.
> Depending on your intention, either some code to verify your example or a custom function with a manual defined gradient.

@boeddeker - @ezyang and I rewrote the doc note and incorporated more background knowledge for complex derivatives in the updated version. you can check out how the html rendering in the PR description.

anjali411 committed:
Update on "Doc note update for complex autograd"


<img width="1679" alt="Screen Shot 2020-10-07 at 1 45 59 PM" src="https://user-images.githubusercontent.com/20081078/95368324-fa7b2d00-08a3-11eb-9066-2e659a4085a2.png">
<img width="1673" alt="Screen Shot 2020-10-07 at 1 46 10 PM" src="https://user-images.githubusercontent.com/20081078/95368332-fbac5a00-08a3-11eb-9be5-77ce6deb8967.png">
<img width="1667" alt="Screen Shot 2020-10-07 at 1 46 30 PM" src="https://user-images.githubusercontent.com/20081078/95368337-fe0eb400-08a3-11eb-80a2-5ad23feeeb83.png">
<img width="1679" alt="Screen Shot 2020-10-07 at 1 46 48 PM" src="https://user-images.githubusercontent.com/20081078/95368345-00710e00-08a4-11eb-96d9-e2d544554a4b.png">
<img width="1680" alt="Screen Shot 2020-10-07 at 1 47 03 PM" src="https://user-images.githubusercontent.com/20081078/95368350-023ad180-08a4-11eb-89b3-f079480741f4.png">
<img width="1680" alt="Screen Shot 2020-10-07 at 1 47 12 PM" src="https://user-images.githubusercontent.com/20081078/95368364-0535c200-08a4-11eb-82fc-9435a046e4ca.png">


[ghstack-poisoned]

anjali411 committed:
Update on "Doc note update for complex autograd"


<img width="1679" alt="Screen Shot 2020-10-07 at 1 45 59 PM" src="https://user-images.githubusercontent.com/20081078/95368324-fa7b2d00-08a3-11eb-9066-2e659a4085a2.png">
<img width="1673" alt="Screen Shot 2020-10-07 at 1 46 10 PM" src="https://user-images.githubusercontent.com/20081078/95368332-fbac5a00-08a3-11eb-9be5-77ce6deb8967.png">
<img width="1667" alt="Screen Shot 2020-10-07 at 1 46 30 PM" src="https://user-images.githubusercontent.com/20081078/95368337-fe0eb400-08a3-11eb-80a2-5ad23feeeb83.png">
<img width="1679" alt="Screen Shot 2020-10-07 at 1 46 48 PM" src="https://user-images.githubusercontent.com/20081078/95368345-00710e00-08a4-11eb-96d9-e2d544554a4b.png">
<img width="1680" alt="Screen Shot 2020-10-07 at 1 47 03 PM" src="https://user-images.githubusercontent.com/20081078/95368350-023ad180-08a4-11eb-89b3-f079480741f4.png">
<img width="1680" alt="Screen Shot 2020-10-07 at 1 47 12 PM" src="https://user-images.githubusercontent.com/20081078/95368364-0535c200-08a4-11eb-82fc-9435a046e4ca.png">


[ghstack-poisoned]

ezyang commented on docs/source/notes/autograd.rst:
```diff
@@ -214,80 +214,250 @@ proper thread locking code to ensure the hooks are thread safe.
-convention for autograd for Complex Numbers.
+- PyTorch Autograd returns gradients that are directly usable for optimization and gradient descent (no conjugation needed),
+  so existing optimizers work out of the box. We don’t make any attempt to support holomorphic functions specially;
+  in particular, derivatives are only well defined for real-valued loss functions.
```
I realize this is my own damn fault because I wrote this text in the first place, but taken literally this text is wrong (I'm not sure what I even meant when I wrote this)

anjali411 replied:
hmmm I rephrased it. does it look any better?

anjali411 committed:
Update on "Doc note update for complex autograd"


<img width="1679" alt="Screen Shot 2020-10-07 at 1 45 59 PM" src="https://user-images.githubusercontent.com/20081078/95368324-fa7b2d00-08a3-11eb-9066-2e659a4085a2.png">
<img width="1673" alt="Screen Shot 2020-10-07 at 1 46 10 PM" src="https://user-images.githubusercontent.com/20081078/95368332-fbac5a00-08a3-11eb-9be5-77ce6deb8967.png">
<img width="1667" alt="Screen Shot 2020-10-07 at 1 46 30 PM" src="https://user-images.githubusercontent.com/20081078/95368337-fe0eb400-08a3-11eb-80a2-5ad23feeeb83.png">
<img width="1679" alt="Screen Shot 2020-10-07 at 1 46 48 PM" src="https://user-images.githubusercontent.com/20081078/95368345-00710e00-08a4-11eb-96d9-e2d544554a4b.png">
<img width="1680" alt="Screen Shot 2020-10-07 at 1 47 03 PM" src="https://user-images.githubusercontent.com/20081078/95368350-023ad180-08a4-11eb-89b3-f079480741f4.png">
<img width="1680" alt="Screen Shot 2020-10-07 at 1 47 12 PM" src="https://user-images.githubusercontent.com/20081078/95368364-0535c200-08a4-11eb-82fc-9435a046e4ca.png">


[ghstack-poisoned]

ezyang committed:
Update on "Doc note update for complex autograd"


<img width="1679" alt="Screen Shot 2020-10-07 at 1 45 59 PM" src="https://user-images.githubusercontent.com/20081078/95368324-fa7b2d00-08a3-11eb-9066-2e659a4085a2.png">
<img width="1673" alt="Screen Shot 2020-10-07 at 1 46 10 PM" src="https://user-images.githubusercontent.com/20081078/95368332-fbac5a00-08a3-11eb-9be5-77ce6deb8967.png">
<img width="1667" alt="Screen Shot 2020-10-07 at 1 46 30 PM" src="https://user-images.githubusercontent.com/20081078/95368337-fe0eb400-08a3-11eb-80a2-5ad23feeeb83.png">
<img width="1679" alt="Screen Shot 2020-10-07 at 1 46 48 PM" src="https://user-images.githubusercontent.com/20081078/95368345-00710e00-08a4-11eb-96d9-e2d544554a4b.png">
<img width="1680" alt="Screen Shot 2020-10-07 at 1 47 03 PM" src="https://user-images.githubusercontent.com/20081078/95368350-023ad180-08a4-11eb-89b3-f079480741f4.png">
<img width="1680" alt="Screen Shot 2020-10-07 at 1 47 12 PM" src="https://user-images.githubusercontent.com/20081078/95368364-0535c200-08a4-11eb-82fc-9435a046e4ca.png">


[ghstack-poisoned]

anjali411 committed:
Update on "Doc note update for complex autograd"


<img width="1679" alt="Screen Shot 2020-10-07 at 1 45 59 PM" src="https://user-images.githubusercontent.com/20081078/95368324-fa7b2d00-08a3-11eb-9066-2e659a4085a2.png">
<img width="1673" alt="Screen Shot 2020-10-07 at 1 46 10 PM" src="https://user-images.githubusercontent.com/20081078/95368332-fbac5a00-08a3-11eb-9be5-77ce6deb8967.png">
<img width="1667" alt="Screen Shot 2020-10-07 at 1 46 30 PM" src="https://user-images.githubusercontent.com/20081078/95368337-fe0eb400-08a3-11eb-80a2-5ad23feeeb83.png">
<img width="1679" alt="Screen Shot 2020-10-07 at 1 46 48 PM" src="https://user-images.githubusercontent.com/20081078/95368345-00710e00-08a4-11eb-96d9-e2d544554a4b.png">
<img width="1680" alt="Screen Shot 2020-10-07 at 1 47 03 PM" src="https://user-images.githubusercontent.com/20081078/95368350-023ad180-08a4-11eb-89b3-f079480741f4.png">
<img width="1680" alt="Screen Shot 2020-10-07 at 1 47 12 PM" src="https://user-images.githubusercontent.com/20081078/95368364-0535c200-08a4-11eb-82fc-9435a046e4ca.png">


[ghstack-poisoned]
